src/OutputWriters/netcdf_output_writer.jl

```diff
@@ -431,25 +1067,173 @@ function NetCDFOutputWriter(model, outputs;
-    default_curvilinear_dimension_attributes
+    # Merge the default with any user-supplied output attributes, ensuring the user-supplied ones
+    # can overwrite the defaults.
+    output_attributes = merge(
+        default_dimension_attributes(grid, dimension_name_generator),
+        default_output_attributes(model),
+        output_attributes
+    )
 
-get_default_dimension_attributes(grid::ImmersedBoundaryGrid) =
-    get_default_dimension_attributes(grid.underlying_grid)
+    # Define variables for each dimension and attributes if this is a new file.
+    if mode == "c"
+        # This metadata is to support `FieldTimeSeries`.
+        write_grid_reconstruction_metadata!(dataset, grid, indices, array_type, deflatelevel)
+
+        # DateTime and TimeDate are both <: AbstractTime
+        time_attrib = model.clock.time isa AbstractTime ?
+            Dict("long_name" => "Time", "units" => "seconds since 2000-01-01 00:00:00") :
+            Dict("long_name" => "Time", "units" => "seconds")
+
+        # Create an unlimited dimension "time"
+        # Time should always be Float64 to be extra safe from rounding errors.
+        # See: https://github.com/CliMA/Oceananigans.jl/issues/3056
+        defDim(dataset, "time", Inf)
+        defVar(dataset, "time", Float64, ("time",), attrib=time_attrib)
+
+        # Create spatial dimensions as variables whose dimensions are themselves.
+        # Each should already have a default attribute.
+        for (dim_name, dim_array) in dims
+            defVar(dataset, dim_name, array_type(dim_array), (dim_name,),
+                   deflatelevel=deflatelevel, attrib=output_attributes[dim_name])
+        end
+
+        time_independent_vars = Dict()
+
+        if include_grid_metrics
+            grid_metrics = gather_grid_metrics(grid, indices, dimension_name_generator)
+            merge!(time_independent_vars, grid_metrics)
+        end
+
+        if grid isa ImmersedBoundaryGrid
+            immersed_boundary_vars = gather_immersed_boundary(grid, indices, dimension_name_generator)
+            merge!(time_independent_vars, immersed_boundary_vars)
+        end
+
+        if !isempty(time_independent_vars)
+            for (name, output) in sort(collect(pairs(time_independent_vars)), by=first)
+                output = construct_output(output, grid, indices, with_halos)
+                attributes = haskey(output_attributes, name) ? output_attributes[name] : Dict()
+                materialized = materialize_output(output, model)
+
+                define_output_variable!(
+                    dataset,
+                    materialized,
+                    name,
+                    array_type,
+                    deflatelevel,
+                    attributes,
+                    dimensions,
+                    filepath, # for better error messages
+                    dimension_name_generator,
+                    false # time_dependent = false
+                )
```

Comment:
```suggestion
                define_output_variable!(dataset,
                                        materialized,
                                        name,
                                        array_type,
                                        deflatelevel,
                                        attributes,
                                        dimensions,
                                        filepath, # for better error messages
                                        dimension_name_generator,
                                        false # time_dependent = false)
```

Reply:
Agree with this comment generally, we can start a discussion to format the code differently throughout, but either way I think we should strive to format consistently.

Reply:
Yes I think I got carried away with the refactoring haha, but 100% agree the code should conform to the existing style. I will format the code consistently.